server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java

```diff
@@ -121,38 +136,94 @@ private CurrentWsResponse toWsResponse(DbSession dbSession, String userLogin) {
       .collect(toList());
   }
 
-  private CurrentWsResponse.Homepage findHomepageFor(DbSession dbSession, UserDto user) {
-    if (user.getHomepageType() == null) {
-      return defaultHomepageOf();
+  private CurrentWsResponse.Homepage buildHomepage(DbSession dbSession, UserDto user) {
+    if (noHomepageSet(user)) {
+      return defaultHomepage();
+    }
+
+    return doBuildHomepage(dbSession, user).orElse(defaultHomepage());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> doBuildHomepage(DbSession dbSession, UserDto user) {
+
+    if (PROJECT.toString().equals(user.getHomepageType())) {
+      return projectHomepage(dbSession, user);
+    }
+
+    if (APPLICATION.toString().equals(user.getHomepageType()) || PORTFOLIO.toString().equals(user.getHomepageType())) {
+      return applicationAndPortfolioHomepage(dbSession, user);
+    }
+
+    if (ORGANIZATION.toString().equals(user.getHomepageType())) {
+      return organizationHomepage(dbSession, user);
+    }
+
+    return of(CurrentWsResponse.Homepage.newBuilder()
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .build());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> projectHomepage(DbSession dbSession, UserDto user) {
+    Optional<ComponentDto> projectOptional = ofNullable(dbClient.componentDao().selectByUuid(dbSession, of(user.getHomepageParameter()).orElse(EMPTY)).orNull());
+    if (shouldCleanProjectHomepage(projectOptional)) {
+      cleanUserHomepageInDb(dbSession, user);
+      return empty();
     }
-    String homepageParameter = getHomepageParameter(dbSession, user.getHomepageType(), user.getHomepageParameter());
+
     CurrentWsResponse.Homepage.Builder homepage = CurrentWsResponse.Homepage.newBuilder()
-      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()));
-    setNullable(homepageParameter, homepage::setParameter);
-    return homepage.build();
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .setComponent(projectOptional.get().getKey());
+    setNullable(projectOptional.get().getBranch(), homepage::setBranch);
+    return of(homepage.build());
   }
 
-  @CheckForNull
-  private String getHomepageParameter(DbSession dbSession, String homepageType, String homepageParameter) {
-    if (PROJECT.toString().equals(homepageType)) {
-      return dbClient.componentDao().selectByUuid(dbSession, homepageParameter)
-        .transform(ComponentDto::getKey)
-        .or(() -> {
-          throw new IllegalStateException(format("Unknown component '%s' for homepageParameter", homepageParameter));
-        });
+  private boolean shouldCleanProjectHomepage(Optional<ComponentDto> projectOptional) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Specify a "ComponentDto" parameter instead. ([squid:S3553](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1185%2Fhomepage_on_sonarqube&issues=AWHdP9XFw6XGHKvPuY57&open=AWHdP9XFw6XGHKvPuY57)

[ISSUE_KEY]: AWHdP9XFw6XGHKvPuY57